tests/Console/ConfigurationResolverTest.php

```diff
@@ -868,6 +869,123 @@ public function testResolveRulesWithOption()
         );
     }
 
+    /**
+     * @param string[] $rules
+     * @param string[] $negativeRules
+     *
+     * @dataProvider provideSubtractionCases
+     */
+    public function testResolveRulesWithNegativeSet(array $rules, array $negativeRules)
```

Comment:
I would not introduce such a big change into this PR.
Still, I do believe that it's possible to move logic outside of `test*` method